### DISCUSSION TITLE
[Bug] Cannot bind class.

### DISCUSSION THREAD

**Discussion by ghost** [wjakob/nanobind#229] [Q&A] [answered]
Hi all ! Hi @wjakob !

I come back to you with an error that I never saw before and I cannot figure out what it means.

I tried with nanobind v1.3.2, v1.4.0 & master.
For the CMake, I tried the `nanobind_add_module` way and the low-level interface way, which I am normally using.

Here is the error I get :

```
      Undefined symbols for architecture arm64:
        "__ZN7COESA76C1Ev", referenced from:
            __ZZN8nanobind6detail11func_createILb0ELb1EZNS_4initIJEE7executeINS_6class_I7COESA76JEEEJEEEvRT_DpRKT0_EUlNS_18pointer_and_handleIS6_EEE_vJSF_EJLm0EEJNS_5scopeENS_4nameENS_9is_methodEEEEP7_objectOT1_PFT2_DpT3_ESt16integer_sequenceImJXspT4_EEEDpRKT5_ENUlPvPSL_PhNS_9rv_policyEPNS0_12cleanup_listEE_4_FUNESZ_S10_S11_S12_S14_ in _ode_solver_bindings.cpp.o
```

which appears only when I add

```
    nb::class_<COESA76>(m, "COESA76")
            .def(nb::init<>());
```

to the module that is otherwise defined like so:

```
NB_MODULE(_ode_solver_kernel, m) {
    m.doc() =
            "Python-C++ bindings for the binary kernel for solving ordinary differential equations "
            "with SUNDIALS.";

    m.def("_solve", _solve, "initial_position"_a, "initial_velocity"_a, "t0"_a, "tf"_a,
          "output_frequency"_a);
}
```

The class itself is defined like so:

```
#include <vector>


class COESA76 {
  public:
    COESA76();
    ~COESA76() = default;
    void _check_altitude(double const alt);
    int _get_index(double x, std::vector<double> const& x_levels);
    double density(double altitude);
    double temperature(double altitude);
    double pressure(double altitude);
}
```

with some other private constants (doubles and vectors). It really does not look like a complex class to me, so I am completely stumped by the error message.

Can someone please advise, I need this bind fairly soon ...

Thank you !!

**Comment by wjakob** [comment]
Can you take a look at the FAQ entry related to symbol visibility and see if that helps?

**Reply by ghost** [reply]
Thanks @wjakob ! 🙂

Do you mean the entry about the multi library project ? 

Could it be that, since I first compile the code in a separate library and then I use `nanobind_add_module`, it is confused about which version of the class to use, the one in the library or the one in the source files ?
In `nanobind_add_module`, should I only input the cpp files that contain bindings ? 🤨

**Reply by wjakob** [reply, marked as answer]
``__ZN7COESA76C1Ev`` is the mangled form of ``COESA76::COESA76()`` (you can use the ``c++filt`` program to de-mangle symbols). So the nanobind extension can't resolve the symbol of the constructor, which is a basic C++ compilation/linking issue and not a nanobind problem.

Likely, your class will need `__attribute__ ((visibility("default")))` annotations to make sure its symbols are actually exported. Or you forgot to link the nanobind extension against the other library containing this constructor.

**Reply by ghost** [reply]
Hey @wjakob, sorry for the delay.

It was not really a visibility problem, it was forgetfulness from my end that nanobind needs the constructor to be set to default manually  and cannot just be left empty, _i.e._ `COESA76() = default;` rather than just `COESA76();`.

Thanks for the support !
